pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -254,7 +256,9 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
   @Override
   protected GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo) {
     assert _partialUpsertHandler != null;
-    AtomicReference<GenericRow> previousRecordReference = new AtomicReference<>();
+    AtomicReference<GenericRow> mergedRowReference = new AtomicReference<>();
```

Comment:
The change can happen in-place, and we don't really need to create this reference. We may directly modify `record`

Reply:
gotcha. i have pushed the logic from the upsertMetadata manager to handler, and do in-place merging in the handler.